### PR TITLE
Thorn automagically fills required fields even when their values have been supplied

### PR DESCRIPTION
**Pull request by jcsmorais**
While creating fixtures, if one supplies a falsy (from a JS perspective) value to a required field, Thorn automagically fills those fields as if no values have been supplied.

E.g.: 
```
Module: Opportunities
Field: amount
Value: 0.0
```
E.g.: 
```javascript
var x = 0.0;
console.log(!x ? 'foo' : 'bar'); // foo is printed
```
The bug seems to be [here](https://github.com/sugarcrm/thorn/blob/master/index.js#L336).

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -333,7 +333,7 @@ let Fixtures = {
             let getRequiredFieldPromise = MetadataHandler.getRequiredFields(model.module)
             .then((requiredFields) => {
                 _.each(requiredFields, (field) => {
-                    if (!request.data[field.name]) {
+                    if (_.isUndefined(request.data[field.name])) {
```
Comment (jcsmorais) [inline index.js]: We should only try to automagically fill required fields if they haven't been supplied.

File: metadata.json [hunk 2]
```diff
@@ -1,8 +1,24 @@
 {
   "TestModule1": {
-    "fields": []
+    "fields": {
+      "testField3": {
```
Comment (jcsmorais) [inline metadata.json]: `testField3` and `testField4` are just a couple of examples of fields that might have falsy values.
`testField5` on the other hand is just here to make sure that it actually gets automagically filled when nothing is supplied.

File: tests/index.js [hunk 3]
```diff
@@ -106,6 +106,8 @@ describe('Thorn', () => {
                         name: 'TestRecord1',
                         testField1: 'TestField1data',
                         testField2: 'TestField2data',
+                        testField3: false,
```
Comment (jcsmorais) [inline tests/index.js]: I was on the fence between adding a new test-case such as `it('should fill required fields when respective values aren\'t supplied' or making use of one of the already existing ones, went with the latter 'cause it made me feel that having a separate test-case for something that is kind of being tested already was overkill. I'm open to suggestions though.

File: tests/index.js [hunk 4]
```diff
@@ -120,7 +122,13 @@ describe('Thorn', () => {
 
                         expect(request.url).to.contain(fixture.module);
                         expect(request.method).to.equal('POST');
-                        expect(request.data).to.eql(fixture.attributes);
```
Comment (jcsmorais) [inline tests/index.js]: One of the things I noticed is that in this kind of tests where we're testing fixtures creation, we're not deeply comparing the values, `request.data`, being sent to the server which in turn might lead to the corner case we're fixing here, but on the other hand it also makes sense to do it this way since the data we're testing is actually the data we care about, whatever is being automagically filled by thorn/faker is considered accessory, which is also why I'm testing each individual data attribute on this test-case and not in the other ones.

File: tests/index.js (hunk 3 above)
Comment (rhoggSugarcrm) [inline tests/index.js]: Since the test file is already pretty big (#142), I'm okay with reusing the existing test.

As a compromise, IMO we should put a comment explaining that we're testing this bug here.

Comment (rhoggSugarcrm) [inline tests/index.js]: It's worth noting that this test doesn't seem to fail if you revert the change [up here](https://github.com/sugarcrm/thorn/pull/134/files#diff-168726dbe96b3ce427e7fedce31bb0bcL336).

So, upon reflection, it might make sense to create a consistently failing test (perhaps by exploiting Sinon to force `Math.random()` to return a certain value).

Comment (rhoggSugarcrm) [inline tests/index.js]: I think I've discovered the reason: the fixtures object is somehow being modified as a reference. If you use `_.clone()` on it, you'll discover that the test fails once you revert the primary change.

Comment (jcsmorais) [inline tests/index.js]: Imho the test does make sense, I'm worried that it might be highlighting a bigger problem than the one I was initially expecting and that now seems to be similar to the one we had on [wrapRequest](https://github.com/sugarcrm/thorn/pull/79/files#r96758868) that has been fixed as part of #79.

File: index.js [hunk 5]
```diff
@@ -320,7 +320,7 @@ let Fixtures = {
             let request = {
                 url: '/' + VERSION + '/' + model.module,
                 method: 'POST',
-                data: model.attributes || {},
+                data: _.cloneDeep(model.attributes) || {},
```
Comment (jcsmorais) [inline index.js]: This portion addresses the problem discussed [here](https://github.com/sugarcrm/thorn/pull/134#discussion_r102377004).

TL;DR

If someone declares a fixture and relies on it after calling `Fixtures.create`, if that fixture has required fields that are not supplied as attributes then the original fixture changes when we automatically populate those required attributes/fields.

Comment (jcsmorais) [inline index.js]: This portion addresses the problem discussed [here](https://github.com/sugarcrm/thorn/pull/134#discussion_r102377004).

TL;DR

If someone declares a fixture and relies on it after calling `Fixtures.create`, if that fixture has required fields that are not supplied as attributes then the original fixture changes when we automatically populate those required attributes/fields.

File: tests/index.js [hunk 6]
```diff
@@ -120,16 +120,19 @@ describe('Thorn', () => {
...
+                        expect(request.data).to.contain(fixture.attributes);
 
                         expect(this.req.headers['x-thorn']).to.equal('Fixtures');
 
                         return constructBulkResponse({
                             _module: fixture.module,
                             id: 'TestId1',
-                            name: fixture.attributes.name,
-                            testField1: fixture.attributes.testField1,
-                            testField2: fixture.attributes.testField2,
```
Comment (jcsmorais) [inline tests/index.js]: Returned values should be based on data retrieved from the request data, not the fixture itself, we specially need to do that for attributes that were not initially defined at the fixture level, e.g: testField[3|4|5]

File: index.js (hunk 5 above)
Comment (jcsmorais) [inline index.js]: In order to make it clearer for everyone, why we need the code above, here's an example that illustrates that:

Remove said code.

Copy-paste the code below [here](https://github.com/sugarcrm/thorn/pull/134/files#diff-4663341696a08eef2606d83e7fb282a9R198)
```
console.log('%j', cmpFixture.attributes);
console.log('%j', falsyFixture.attributes);
console.log('%j', request.data);
```

Run the tests and you'll get something like:
```
{"name":"TestRecord1","testField1":"TestField1data","testField2":"TestField2data","testField3":false,"testField4":0}
{"name":"TestRecord1","testField1":"TestField1data","testField2":"TestField2data","testField3":false,"testField4":0,"testField5":"01z64fu6di8tzldnr5r05wj0p7wctv"}
{"name":"TestRecord1","testField1":"TestField1data","testField2":"TestField2data","testField3":false,"testField4":0,"testField5":"01z64fu6di8tzldnr5r05wj0p7wctv"}
```

Note that the two objects at the bottom are exactly the same.

Comment (jcsmorais) [inline index.js]: Thinking a little more about it, maybe I'll split the [new test](https://github.com/sugarcrm/thorn/pull/134/files#diff-4663341696a08eef2606d83e7fb282a9R183) into two: 
1. it should only fill required fields when respective values aren't supplied
2. it should not update given fixtures if required fields aren't supplied

I'll do it tomorrow morning and I'll solve the conflicts with the latest changes that got in today.

File: tests/index.js [hunk 7]
```diff
@@ -174,6 +180,53 @@ describe('Thorn', () => {
...
+                        expect(request.method).to.equal('POST');
+
+                        expect(request.data.name).to.equal(cmpFixture.attributes.name);
+                        expect(request.data.testField1).to.equal(cmpFixture.attributes.testField1);
+                        expect(request.data.testField2).to.equal(cmpFixture.attributes.testField2);
+                        expect(request.data.testField3).to.equal(cmpFixture.attributes.testField3);
+                        expect(request.data.testField4).to.equal(cmpFixture.attributes.testField4);
+                        expect(request.data.testField5).to.not.be.undefined;
+
+                        expect(this.req.headers['x-thorn']).to.equal('Fixtures');
```
Comment (rhoggSugarcrm) [inline tests/index.js]: IMO there's no need to test this here; we're testing it elsewhere and it's not material to what this test is about.
